spirv_reflect.c

```diff
@@ -3580,6 +3750,282 @@ static SpvReflectResult ParseExecutionModes(
+    || ((op_code) == SpvOpConstantPipeStorage) || ((op_code) == SpvOpSpecConstantCompositeContinuedINTEL)\
+    || ((op_code) == SpvOpConstantFunctionPointerINTEL) || ((op_code) == SpvOpConstantCompositeContinuedINTEL))
+
+#if SPIRV_REFLECT_ENABLE_CONSTANT_EVALUATION
+#ifdef _MSC_VER
+#define SPIRV_REFLECT_FORCEINLINE __forceinline
+#else
+#define SPIRV_REFLECT_FORCEINLINE __attribute__((always_inline)) inline
+#endif
+
+static SpvReflectResult SPIRV_REFLECT_FORCEINLINE EvaluateResult_Impl(SpvReflectEvaluation* p_eval, SpvReflectPrvEvaluationNode* p_node, SpvReflectEvaluationMode evaluation_mode, SpvReflectPrvParser* p_parser);
+
+// only need node for evaluation.. Type information should not be freed before evaluating for now.
+// may move type info into instruction info later, but now it seems not needed...
+static SpvReflectResult EvaluateResult_Do(SpvReflectPrvEvaluationNode* p_node)
+{
+  return EvaluateResult_Impl(NULL, p_node, SPIRV_REFLECT_EVALUATION_MODE_NORMAL, NULL);
+}
+// 1 PASS parser needs to read specOp and check instruction size
+static SpvReflectResult EvaluateResult_1PASS(SpvReflectEvaluation* p_eval, SpvReflectPrvEvaluationNode* p_node, SpvReflectPrvParser* p_parser)
+{
+  return EvaluateResult_Impl(p_eval, p_node, SPIRV_REFLECT_EVALUATION_MODE_AST_1PASS, p_parser);
+}
+
+// 2 PASS needs parser to get id operands and literals
+static SpvReflectResult EvaluateResult_2PASS(SpvReflectEvaluation* p_eval, SpvReflectPrvEvaluationNode* p_node, SpvReflectPrvParser* p_parser)
+{
+  return EvaluateResult_Impl(p_eval, p_node, SPIRV_REFLECT_EVALUATION_MODE_AST_2PASS, p_parser);
+}
+
+#endif
+
+static SpvReflectResult ParseSpecializationConstants(SpvReflectPrvParser* p_parser, SpvReflectShaderModule* p_module)
+{
+  p_module->specialization_constant_count = 0;
+  p_module->specialization_constants = NULL;
+
+  uint32_t constant_instruction_num = 0;
+  for (size_t i = 0; i < p_parser->node_count; ++i) {
+    SpvReflectPrvNode* p_node = &(p_parser->nodes[i]);
+    // constant types need to be tracked after parser is dead...
+    if (IS_CONSTANT_OP(p_node->op)) {
+        constant_instruction_num++;
+    }
+    if (p_node->op == SpvOpSpecConstantTrue || p_node->op == SpvOpSpecConstantFalse || p_node->op == SpvOpSpecConstant) {
+      p_module->specialization_constant_count++;
+    }
+  }
+
+  if (p_module->specialization_constant_count == 0) {
+    return SPV_REFLECT_RESULT_SUCCESS;
+  }
+
+  p_module->specialization_constants = (SpvReflectSpecializationConstant*)calloc(p_module->specialization_constant_count, sizeof(SpvReflectSpecializationConstant));
+  if(IsNull(p_module->specialization_constants)){
+    return SPV_REFLECT_RESULT_ERROR_ALLOC_FAILED;
+  }
+
+  // deal with WorkGroupSize builtin
+  if (p_module->_internal->builtin_WorkGroupSize != INVALID_VALUE) {
+    SpvReflectPrvNode* p_node = FindNode(p_parser, p_module->_internal->builtin_WorkGroupSize);
```

Comment:
What is this code block doing? Why is it overwriting the x value of the local size with the OpID that WorkGroupSize is held in?